### PR TITLE
Add SSO support

### PR DESCRIPTION
**Pull request by arturo-seijas**

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -242,12 +245,62 @@ def _get_indico_env_config(self):
...
+                }
+            }
+            env_config["INDICO_IDENTITY_PROVIDERS"] = str(identity_providers)
         env_config["INDICO_EXTRA_PLUGINS"] = ",".join(indico_plugins)
         return env_config
 
+    def _is_config_valid(self):
+        return (
+            not self.config["saml_target_url"] or UBUNTU_SAML_URL == self.config["saml_target_url"]
+        )
```
Comment (mthaddon) [inline src/charm.py:291]: This function currently only checks if the saml_target_url is valid. I think we should either change this to be specifically about the saml config (`is_saml_target_url_valid`) or update it to be extendable to other config options in the future (likely by passing the message to block on).

File: src/charm.py [hunk 2]
```diff
@@ -14,12 +14,13 @@
...
 from ops.main import main
-from ops.model import ActiveStatus, MaintenanceStatus, WaitingStatus
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, WaitingStatus
 from ops.pebble import ExecError
 
 DATABASE_NAME = "indico"
-PORT = 8080
 INDICO_CUSTOMIZATION_DIR = "/srv/indico/custom"
+PORT = 8080
+UBUNTU_SAML_URL = "https://login.ubuntu.com/saml/"
```
Comment (mthaddon) [inline src/charm.py:23]: This is a simple string variable currently, but if we want to support other providers in the future we'd need different variables for each.

File: src/charm.py [hunk 3]
```diff
@@ -242,12 +245,62 @@ def _get_indico_env_config(self):
...
+                    "trusted_email": True,
+                    "mapping": {
+                        "user_name": "username",
+                        "first_name": "fullname",
+                        "last_name": "",
+                        "email": "email",
+                    },
+                    "identifier_field": "openid",
+                }
+            }
```
Comment (mthaddon) [inline src/charm.py:283]: How would these data structures support other providers in the future? Would we set different `auth_providers` and `identity_providers` variables based on which provider(s) the user wanted to support? Do we expect to only ever support one provider, or to support multiple providers?

**Review by mthaddon — CHANGES_REQUESTED** [review]
Some comments inline. I think there's a balance here between making things too complicated now to support a theoretical future where we support multiple providers, and allowing us to add those in the future without needing to make too invasive changes.

One other question - will this _force_ logins via the configured provider, or will users still be able to create local accounts (we'd very much prefer it to be forced)?

File: src/charm.py (hunk 2 above)
Comment (arturo-seijas) [reply src/charm.py:23]: Yes, but I think is good enough for now. YAGNI maybe?

File: src/charm.py (hunk 1 above)
Comment (arturo-seijas) [reply src/charm.py:292]: Changed

**Comment by arturo-seijas** [comment]
> Some comments inline. I think there's a balance here between making things too complicated now to support a theoretical future where we support multiple providers, and allowing us to add those in the future without needing to make too invasive changes.
> 
> One other question - will this _force_ logins via the configured provider, or will users still be able to create local accounts (we'd very much prefer it to be forced)?

An account will be created inside Indico and will be linked to the auth provider. Unfortunately, Indicoo accounts require both name and surname and our SSO service is storing that as a single value, so a little manual set up of the Indico account will be required. When logged in, a prepopulated form will appear to the user to complete the Indico account creation and once that is set up, the existence of this local account will be transparent to the user

**Review by mthaddon — APPROVED** [review]
LGTM

**Pull request merged**